pkg/rook/purge_osd.go

```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2023 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rook
+
+import (
+	ctx "context"
+	"fmt"
+	"log"
+
+	"github.com/rook/kubectl-rook-ceph/pkg/bash"
+	"github.com/rook/kubectl-rook-ceph/pkg/k8sutil"
+	"github.com/rook/kubectl-rook-ceph/pkg/mons"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func PurgeOsd(context *k8sutil.Context, operatorNamespace, clusterNamespace, osdId, flag string) string {
+
+	monCm, err := context.Clientset.CoreV1().ConfigMaps(clusterNamespace).Get(ctx.TODO(), mons.MonConfigMap, v1.GetOptions{})
+	if err != nil {
+		log.Fatalf("failed to get mon configmap %s %v", mons.MonConfigMap, err)
+	}
+
+	monEndPoint := monCm.Data["data"]
+	cephArgs := []string{
+		"auth", "print-key", "client.admin",
+	}
+	adminKey := bash.RunCommandInOperatorPod(context, "ceph", cephArgs, operatorNamespace, clusterNamespace)
+
+	cmd := "/bin/sh"
+	args := []string{
+		"-c",
+		fmt.Sprintf("export ROOK_MON_ENDPOINTS=%s ROOK_CEPH_USERNAME=client.admin ROOK_CEPH_SECRET=%s ROOK_CONFIG_DIR=/var/lib/rook && rook ceph osd remove --osd-ids=%s --force-osd-removal=%s", monEndPoint, adminKey, osdId, flag),
```

Comment:
All of this command has the potential to be seen in plain text so we need to make sure the secrets are in a file path.

Reply:
I tried reading from the file following the above approach that didn't work. Are you suggesting any other way?

I didn't see the secret printing in the logs.

Reply:
```yaml
023-02-20 10:54:25.999916 W | cephcmd: loaded admin secret from env var ROOK_CEPH_SECRET instead of from file
2023-02-20 10:54:26.001142 I | rookcmd: starting Rook v1.11.0-alpha.0.20.g10abbd0d8 with arguments 'rook ceph osd remove --osd-ids=0 --force-osd-removal=true'
```

Reply:
I'm thinking of these vars that the operator uses. This plugin could assume the files exist (depending on the namespace), then no need to look up the secret value.
```
--cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring
```

Reply:
those args are used for ceph command, here we are running rook command that triggers the osd removal

Reply:
While the secret can be in an env var, the intention with #11434 is that callers will move to the new approach of using a file for the secret, to avoid the command line being exposed with the secret. As seen in that PR, the osd purge job will mount the secret to the path `/var/lib/rook-ceph-mon`, which will be read from that path [here](https://github.com/rook/rook/blob/release-1.10/cmd/rook/ceph/osd.go#L201). I wonder if this plugin could first run a command to copy the file into that path so the purge command will then just find it there.

Reply:
Although a complication is that this would require at least Rook v1.10. So if the plugin is targeting an older version of rook, that wouldn't work. We should figure out how to target different versions of rook (separate from this PR).

Reply:
As discussed, let's leave it this way for now and consider a future update based on version.